### DISCUSSION TITLE
Long-Term Roadmap Discussion

### DISCUSSION THREAD

**Discussion by chsh2** [chsh2/nijiGPen#12] [Ideas]
Thanks to all of you for supporting the development of this add-on. So far the basic functionality of several major modules has been implemented. However, there are still improvements and some new modules not yet finished.

This is my amateur project and my artistic skills are also quite limited. Development may take a long time and my understanding and solution may not always be correct. That is why I try to write down these ideas, both as a reminder and for discussion. Any feedback or suggestions, either on existing functions or new ideas are appreciated.

## Motivation & Scope

Grease Pencil is quite powerful and this add-on cannot cover all use cases. The focus of this add-on is to develop features that are useful for creating 2D artwork, including drawing and graphic design. I can think of the following categories of features:

- Common algorithms for 2D curves, but not applicable to 3D curves. Boolean operations are a good example. 
- Features that help 2D artists benefit from Blender's 3D capabilities, because 3D-assisted painting is becoming increasingly popular and I think Blender is a very good tool for it.
- Exchange images and data between Grease Pencil and other popular drawing applications, including both raster-based (e.g. Krita, Photoshop, Clip Studio Paint) and vector-based (e.g. Inkscape, Illustrator).

As for the algorithms to be used, most of the finished features are based on external scientific computing libraries. This is also expected to be the case for most future features. Therefore, I am not currently seeking the opportunity to be integrated into the official Blender release. However, it has been a problem for some users to install the dependencies, which is a problem to be solved.

As for the interface, all functions are now accessed via buttons. I am considering adding more options for convenience, such as modal operators and workspace tools. But I want the add-on to be as unobtrusive as possible, which means that all functions are triggered manually by the user and do not start without the user's knowledge. So I tend to avoid using handlers.

## Functions to Develop

These are some possible new features and improvements I have in mind. This list may be subject to change in the future.

### Stylized Shaders for Mesh Generation

In many cases of meshes generated from Grease Pencil strokes, stylized shaders may look better than PBR ones. It will be more convenient if the add-on can come with some shaders applied to the meshes quickly. Toon shaders and brush simulation shaders might be suitable here.

### From Normal Map to Real 3D Structure

The normal map has its limitations. If we can generate similar geometry in 3D, it will have better shading and more natural transitions between different meshes. 
A good reference is [MonsterMash](https://monstermash.zone/). Also, I have previously implemented a [script](https://github.com/Chaosinism/NormalMapTool) to convert a normal map to depth. I am thinking about whether I can modify it and reuse it here.

### Animated Mesh

It would be useful to be able to generate meshes for multiple frames to create an animation. I know there is another add-on in development, [Claymotion](http://www.cgchannel.com/2022/03/check-out-blenders-upcoming-claymation-add-on/), with a similar idea, but I am not sure of its current status. To avoid duplication of effort, I may need to think of a different way of animating.

### Multi-Layer Image Export

Artists often use Photoshop or similar software to post-process Blender-rendered images. Preserving the layer structure of the Grease Pencil object during rendering would simplify the process. The difficulty is that while PSD is the most popular multi-layer file format, it is not designed for data exchange and is difficult to parse. The existing Python libraries focus more on reading and have limitations when writing PSDs. My current solution is to either implement it myself or call GIMP's batch mode.

### Brush Import

Although raster and vector brushes work in different ways, with so many raster brush resources available online, it would be helpful if we could just extract the texture images from them and use them in Blender. 
Again, the problem is the Adobe format. Most brushes available online are ABR rather than open formats like GBR. It is simpler than PSD but has no official documentation as far as I know. Maybe the references are still GIMP and Krita.

### Colored Image Processing

Both the default image tracing and the line art tracing in the add-on work best with monochrome images. I do not know if there is a good way to handle colored raster images.
Some state-of-the-art methods simply quantize the colours and trace each of them. But since Blender represents holes differently than SVG (with holdout materials), I am not sure if such methods will work in Blender without changes.
Another idea I have is to stylize the image with brushes, like the effects in [FotoSketcher](https://fotosketcher.com). 

### Working Plane Detection

It is a bit rigid that current methods only work correctly in three orthogonal 2D planes. Perhaps more options can be added, such as the screen view plane and automatic estimation of the 2D plane based on selected points, which can be useful when drawing on the surface of 3D objects.

### Documentation

Formal documentation should be produced to explain each operator and its parameters. Also, since 2D artists may not be familiar with Blender's 3D environment settings, it may be helpful to have some tips and tutorials on how to set up a scene.

**Comment by b4zz4** [comment]
Some programs that pass from GIMP to SIFZ (easy to modify) and COA from gimp to blender:
 
- https://github.com/d-j-a-y/Gimp2Synfig
- https://github.com/ndee85/coa_tools

Lumo: Beautiful document that describes how to generate lights and shadows for 2D

- http://ivizlab.sfu.ca/arya/Papers/ACM/NPAR-02/Illumination%20for%20Cel%20Animation.pdf

Similar to MonsterMash in blender:

- https://github.com/FujiSunflower/love2d3d
